### DISCUSSION TITLE
Apollo Initialization with next.js locale

### DISCUSSION THREAD

**Discussion by marcin-piechaczek** [vercel/next.js#19261] [Help]
Hey guys, I'm using [this example](https://github.com/vercel/next.js/blob/canary/examples/with-apollo-and-redux/lib/apollo.js) for my app. In my case i need to pass a user browser language as an additional header parameter like this:

```js
   // lib/apollo.js
   ...
  return new ApolloClient({
    ssrMode: !process.browser,
    link: new HttpLink({
      uri,
      credentials: 'include',
      headers: {
        authorization: token ? `Bearer ${token}` : '',
        store: locale // user browser locale, this will be dynamic when user change language
      }
    }),
   ...
```

What is the best way to pass the current language so that it is visible during apollo initialization?

I tried to pass it in _app.js, but it actually works badly:

```js
// pages/_app.js
function MyApp({ Component, pageProps }) {
  const router = useRouter();
  const { locale, defaultLocale } = router;
  const store = useStore(pageProps.initialReduxState);
  const apolloClient = useApollo(pageProps.initialApolloState, locale);

  return (
    <Provider store={store}>
      <IntlProvider locale={locale} defaultLocale={defaultLocale} messages={messages[locale]}>
        <ApolloProvider client={apolloClient}>
          <GlobalStyle />
          <ThemeProvider theme={theme}>
            <Layout>
              <Component {...pageProps} />
            </Layout>
          </ThemeProvider>
        </ApolloProvider>
      </IntlProvider>
    </Provider>
  );
}

export default MyApp;
```

Is this approach correct? Or should I pass the additional header at the query level directly into the components? Any clarifications are welcome.

**Comment by bretthadley** [comment]
Did you find answer to this? I same issue with custom headers for apollo initialization. Where does 'token' come from in your example?

**Reply by marcin-piechaczek** [reply]
You can use localStorage, cookie or [apollo reactive variables](https://www.apollographql.com/docs/react/local-state/reactive-variables/). In my case mix of apollo reactive variables + cookie do the trick. The hard part here is that your are on both sides (browser, server) so if you are going with the localStorage please make sure to add additional if statement to check if you are on the browser side.

**Comment by Aztriltus** [comment]
Reviving an old topic here. I managed to use the Router instance to grab the locale and add it to the apollo client.

```javascript
// apollo-client.ts
import { ApolloClient, InMemoryCache, HttpLink, from } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import Router from 'next/router'

const link = from([
  errorLink,
  new HttpLink({ uri: process.env.API_URL }),
])

const addHeadersLink = setContext((_, { headers }) => {
  return {
    headers: {
      ...headers,
      ...(Router.locale && { locale: Router.locale }),
    },
  }
})

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: addHeadersLink.concat(link),
})

export default client
```

**Reply by dbenfouzari** [reply]
@Aztriltus with your solution I get an error
```sh
networkError: Error: No router instance found.
You should only use "next/router" on the client side of your app.
```

This may be cause I use `getStaticProps` to fetch data.
